Feature/wincher modal improvements

KaisZaoualiWincher opened the pull request:
## Context
<!--
What do we want to achieve with this PR? Why did we write this code?
-->

Add improvements to Wincher modal in the post edit page

## Summary

<!--
Attach one of the following labels to the PR: `changelog: bugfix`, `changelog: enhancement`, `changelog: other`, `changelog: non-user-facing`.
If the changelog item is a bugfix, please use the following sentence structure: Fixes a bug where ... would ... (when ...).
If the changelog item is meant for the changelog of another add-on, start your changelog item with the name of that add-on's repo between square brackets, for example: * [wordpress-seo-premium] Fixes a bug where ....
If the changelog items is meant for the changelog of a javascript package, specify between square brackets in which package changelog the item should be included, for example: * [@yoast/components] Fixes a bug where ....
If the same changelog item is applicable to multiple changelogs/add-ons, add a separate changelog item for all of them.
-->
This PR can be summarized in the following changelog entry:

- Add period picker (Last day, Last week, Last month (default), Last 3 months, Last year. This needs to be based on the history limit for the account, i.e. only show the options applicable to the limit.
- Add an expandable chart
- Add the last updated column
- Add the change value + caret

## Relevant technical choices:

*

## Test instructions
<!--
Please follow these guidelines when creating test instructions:
- Please provide step-by-step instructions how to reproduce the issue, if applicable.
- Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->
### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

* Check account history days limit via "app.wincher.com" (`account` endpoint)
![image](https://github.com/Yoast/wordpress-seo/assets/129404933/fcbb5b80-dac9-4e9a-be5f-b69e5cff973d)
* Make sure that the period picker returns options that respect the history days limit e.g. "last day", "last week" and "last month" for a limit of 31 days.
* Make sure that we are displaying data in the table and in the chart that is using the selected period option.
* Make sure we can select only tracked keyphrases.
* Make sure that "select all" selects all tracked keyphrases.
* Make sure that "deselect all" deselects all selected keyphrases.
* Make sure that we can show chart only if there is a selected keyphrases.
* Make sure that deselecting all keyphrases hide the chart.
* Make sure we display the correct position change, with the correct caret type
* Make sure untracking a keyphrase deselects it.

#### Relevant test scenarios
* [ ] Changes should be tested with the browser console open
* [ ] Changes should be tested on different posts/pages/taxonomies/custom post types/custom taxonomies
* [ ] Changes should be tested on different editors (Block/Classic/Elementor/other)
* [ ] Changes should be tested on different browsers
* [ ] Changes should be tested on multisite
<!--
If you have checked any of the above cases, please add some context about the reason, what to check in the console,
which type/editor/browser should be tested in particular, multisite with subfolders or subdomains, etc.
-->

### Test instructions for QA when the code is in the RC
<!--
Sometimes some steps from the test instructions for the acceptance test aren't relevant anymore once the code has been merged or the feature is complete. If that is the case, do not check the checkbox below.
QA is our Quality Assurance team. The RC is the release candidate zip that is tested before a release
-->

* [x] QA should use the same steps as above.

<!--
If the above checkbox has not been checked, write down all steps QA should take to test this PR, not only the difference with the acceptance test steps. If QA should use the test instructions specified on the epic, paste a link to the relevant comment on the epic.
-->
QA can test this PR by following these steps:

*

## Impact check
<!--
Sometimes PRs have a bigger impact than is suggested in the user-facing changes. In such cases,
additional (regression) testing might be necessary. To make it clear what parts might need additional testing, please outline which parts of the plugin have been impacted by this PR.
-->
This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Other environments

* [ ] This PR also affects Shopify. I have added a changelog entry starting with `[shopify-seo]`, added test instructions for Shopify and attached the `Shopify` label to this PR.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities.
* [x] During testing, I had activated [all plugins that Yoast SEO provides integrations for](https://github.com/Yoast/wordpress-seo/blob/trunk/readme.txt#L106).
* [ ] I have added unit tests to verify the code works as intended.
* [ ] If any part of the code is behind a feature flag, my test instructions also cover cases where the feature flag is switched off.
* [x] I have written this PR in accordance with my team's definition of done.
* [x] I have checked that the base branch is correctly set.

## Innovation

* [ ] No innovation project is applicable for this PR.
* [ ] This PR falls under an innovation project. I have attached the `innovation` label.
* [ ] I have added my hours to [the WBSO document](http://yoa.st/wbso).

Fixes #

## Evidence

![image](https://github.com/Yoast/wordpress-seo/assets/129404933/ac66a0f0-d8fe-4e51-970d-6d93bfc7f874)

KaisZaoualiWincher committed:
Add updated at to wincher seo performance table

KaisZaoualiWincher committed:
Add position change to wincher seo performance table

KaisZaoualiWincher committed:
Add support of start at for keyphrases data

KaisZaoualiWincher committed:
Add chart

KaisZaoualiWincher committed:
Add wincher ranking history

KaisZaoualiWincher committed:
Fix test and linting

skrivanos commented:
@KaisZaoualiWincher Think we need to update trunk in this repo if this is based on trunk in the upstream?

KaisZaoualiWincher commented:
> @KaisZaoualiWincher Think we need to update trunk in this repo if this is based on trunk in the upstream?

Done, and branch rebased :)

skrivanos commented on packages/js/package.json:
```diff
@@ -51,22 +56,20 @@
     "redux-thunk": "^2.2.0",
     "styled-components": "^5.3.6",
     "yoast-components": "^5.24.0",
     "yoastseo": "^1.91.1",
```
A lot of changes to this file and new libs that I don't think are relevant to this branch?

KaisZaoualiWincher replied:
Those are not newly added packages. It's the lerna/yarn add that fixes the package's order if it's incorrect, and those were not sorted correctly previously. Github doesn't make it clear when it's about a line change, maybe better to use vscode for review, it's more clear. ;)

KaisZaoualiWincher replied:
![image](https://github.com/wincher-ab/wordpress-seo/assets/129404933/84d87d6b-f881-4bf4-be10-cfc093941586)

skrivanos commented on packages/js/src/components/WincherPerformanceReport.js:
```diff
@@ -16,7 +16,7 @@ import { Alert, NewButton } from "@yoast/components";
 import { getKeyphrasePosition, PositionOverTimeChart } from "./WincherTableRow";
 import WincherReconnectAlert from "./modals/WincherReconnectAlert";
-import WincherUpgradeCallout from "./modals/WincherUpgradeCallout";
+import WincherUpgradeCallout, { useTrackingInfo } from "./modals/WincherUpgradeCallout";
```
Seems sorta strange we're importing `useTrackingInfo` from `WincherUpgradeCallout` only to call it and pass it back into the `WincherUpgradeCallout`

KaisZaoualiWincher replied:
We need it outisde `WincherUpgradeCallou` since we use it in `WincherSeoPerformance` for both `WincherUpgradeCallou` (line 486) and `TableContent` (line 500).

skrivanos commented on packages/js/src/components/WincherSEOPerformance.js:
```diff
@@ -251,20 +371,95 @@ const TableContent = ( props ) => {
+		opt => moment( opt.value ).isSameOrAfter( historyLimitDate )
+	);
+
+	const defaultPeriod = orderBy( periodOptions, opt => opt.defaultIndex, "desc" )[ 0 ];
```
Isn't this sorting unnecessary? Can't we just assume the constant is in the correct order? Like add a comment above it to future devs.

KaisZaoualiWincher replied:
Why do we rely on the future dev to be careful about it when we can already make the risk nil 😄

skrivanos commented:
Seems like there are some failing code style checks

KaisZaoualiWincher commented:
> Seems like there are some failing code style checks

Tests and CS should be fixed.

skrivanos commented:
Getting a spinner forever:

<img width="584" alt="Screenshot 2023-07-06 at 16 49 39" src="https://github.com/wincher-ab/wordpress-seo/assets/87463/c6ecf38f-8687-40d0-aa28-33440d591ec8">

response is

```json
{
    "status": 200,
    "results": {}
}
```

skrivanos commented:
This endpoint almost always called twice when opening:

<img width="1157" alt="Screenshot 2023-07-06 at 16 51 10" src="https://github.com/wincher-ab/wordpress-seo/assets/87463/58e57611-0299-4669-ba20-9192ee481675">

KaisZaoualiWincher commented:
> Getting a spinner forever:
> 
> <img alt="Screenshot 2023-07-06 at 16 49 39" width="584" src="https://user-images.githubusercontent.com/87463/251487833-c6ecf38f-8687-40d0-aa28-33440d591ec8.png">
> response is
> 
> ```json
> {
>     "status": 200,
>     "results": {}
> }
> ```

Can't reproduce. What's the request payload?

KaisZaoualiWincher commented:
> This endpoint almost always called twice when opening:
> 
> <img alt="Screenshot 2023-07-06 at 16 51 10" width="1157" src="https://user-images.githubusercontent.com/87463/251488251-58e57611-0299-4669-ba20-9192ee481675.png">

Good catch! there is a request invoked before setting up the start at param

KaisZaoualiWincher committed:
Fix extra Wincher keywords api call

skrivanos commented:
Think we're missing the click action on the mini chart in the table.

skrivanos commented:
> > Getting a spinner forever:
> > <img alt="Screenshot 2023-07-06 at 16 49 39" width="584" src="https://user-images.githubusercontent.com/87463/251487833-c6ecf38f-8687-40d0-aa28-33440d591ec8.png">
> > response is
> > ```json
> > {
> >     "status": 200,
> >     "results": {}
> > }
> > ```
> 
> Can't reproduce. What's the request payload?

Can't repro now either...

skrivanos commented:
There is no tooltip in chart. Can we add one with date and value?

skrivanos commented:
IDK if unchecking should hide the chart? Feels weird. I feel like what's shown in the chart should be reflected by what was checked when you clicked the button.

KaisZaoualiWincher commented:
> IDK if unchecking should hide the chart? Feels weird. I feel like what's shown in the chart should be reflected by what was checked when you clicked the button.

So we shouldn't be able to uncheck / check if chart is displayed?

KaisZaoualiWincher commented:
> Think we're missing the click action on the mini chart in the table.

Didn't notice it. I'll add it.

KaisZaoualiWincher commented:
> > IDK if unchecking should hide the chart? Feels weird. I feel like what's shown in the chart should be reflected by what was checked when you clicked the button.
> 
> So we shouldn't be able to uncheck / check if chart is displayed?

I think the way it is now is more interactive.

skrivanos commented:
> > IDK if unchecking should hide the chart? Feels weird. I feel like what's shown in the chart should be reflected by what was checked when you clicked the button.
> 
> So we shouldn't be able to uncheck / check if chart is displayed?

Yes you should, but changes should only apply to the chart when clicking the button again

skrivanos commented:
> > > IDK if unchecking should hide the chart? Feels weird. I feel like what's shown in the chart should be reflected by what was checked when you clicked the button.
> > 
> > 
> > So we shouldn't be able to uncheck / check if chart is displayed?
> 
> I think the way it is now is more interactive.

It's rather confusing and essentially makes the button serve no purpose.

KaisZaoualiWincher commented:
What about the deselect-all checkbox, now? Should we make an exception for it? Otherwise, it'll be useless. 

![image](https://github.com/wincher-ab/wordpress-seo/assets/129404933/103b6513-514f-4060-9f00-007b278cb3c9)

skrivanos commented:
> What about the deselect-all checkbox, now? Should we make an exception for it? Otherwise, it'll be useless.

I don't understand why it's useless. If you have more than one keyword it has a purpose?

skrivanos commented:
I checked the design and it seems like you're correct here, so we can leave it.

KaisZaoualiWincher committed:
Add tooltip to Wincher ranking history chart

KaisZaoualiWincher committed:
Add mini chart click action

skrivanos commented:
@KaisZaoualiWincher Is tooltip shared between series? I don't have yoast premium atm so can't test

skrivanos commented:
Maybe we should display an actual point when hovering as well?

KaisZaoualiWincher commented:
> @KaisZaoualiWincher Is tooltip shared between series? I don't have yoast premium atm so can't test

Why do you need Yoast premium? You can add another keyword after loading the first keyword rankings and you get both of them in the modal.

KaisZaoualiWincher commented:
> @KaisZaoualiWincher Is tooltip shared between series? I don't have yoast premium atm so can't test

It's not shared since we need `mode = "nearest"` otherwise it'll be hard to trigger the tooltip. I'd like to have a mix of "nearest" and "x" modes but we don't have that.

skrivanos commented:
> > @KaisZaoualiWincher Is tooltip shared between series? I don't have yoast premium atm so can't test
> 
> Why do you need Yoast premium? You can add another keyword after loading the first keyword rankings and you get both of them in the modal.

It's not possible to have more than one focus keyphrase per page without yoast premium...

KaisZaoualiWincher commented:
> > > @KaisZaoualiWincher Is tooltip shared between series? I don't have yoast premium atm so can't test
> > 
> > 
> > Why do you need Yoast premium? You can add another keyword after loading the first keyword rankings and you get both of them in the modal.
> 
> It's not possible to have more than one focus keyphrase per page without yoast premium...

Yes, but you can have more than one keyphrase in the modal if you update the keyphrase:

![image](https://github.com/wincher-ab/wordpress-seo/assets/129404933/db464a0c-67fc-4f9c-a2b1-e452cc579147)

skrivanos commented:
> > @KaisZaoualiWincher Is tooltip shared between series? I don't have yoast premium atm so can't test
> 
> It's not shared since we need `mode = "nearest"` otherwise it'll be hard to trigger the tooltip. I'd like to have a mix of "nearest" and "x" modes but we don't have that.

IMO this is better even though it's not ideal.

```diff
diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
index e4895718fd..9bf04de788 100644
--- a/packages/js/src/components/WincherRankingHistoryChart.js
+++ b/packages/js/src/components/WincherRankingHistoryChart.js
@@ -27,8 +27,8 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
                        y: value,
                } ) ),
                lineTension: 0,
-               pointRadius: 0,
-               pointHoverRadius: 2,
+               pointRadius: 1,
+               pointHoverRadius: 4,
                borderWidth: 2,
        } ) );

@@ -58,7 +58,8 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
                                                },
                                                titleAlign: "center",
                                                intersect: false,
-                                               mode: "nearest",
+                                               mode: "point",
+                                               position: "nearest",
                                        },
                                },
                                scales: {
```

skrivanos commented on packages/js/src/components/WincherRankingHistoryChart.js:
```diff
@@ -0,0 +1,109 @@
+					tooltip: {
+						enabled: true,
+						callbacks: {
+							title: ( x ) => moment( x[ 0 ].label ).format( "YYYY-MM-DD" ),
```
I we need to ensure this prints the date in GMT as otherwise it'll be different from the Wincher UI. Well, in reality it should use the account's time zone I guess `Account.timezone`

Not sure if we should bother syncing these two UIs or not.

KaisZaoualiWincher replied:
> I we need to ensure this prints the date in GMT as otherwise it'll be different from the Wincher UI. Well, in reality it should use the account's time zone I guess `Account.timezone`
> 
> Not sure if we should bother syncing these two UIs or not.

Me neither.

skrivanos replied:
Can you confirm how it works now? Does it print based on local tz?

KaisZaoualiWincher replied:
Yes it displays date + 2AM for me for instance

skrivanos replied:
I think best would be to make it GMT for now..?

KaisZaoualiWincher replied:
Should be fixed

KaisZaoualiWincher commented:
> > > @KaisZaoualiWincher Is tooltip shared between series? I don't have yoast premium atm so can't test
> > 
> > 
> > It's not shared since we need `mode = "nearest"` otherwise it'll be hard to trigger the tooltip. I'd like to have a mix of "nearest" and "x" modes but we don't have that.
> 
> IMO this is better even though it's not ideal.
> 
> ```diff
> diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
> index e4895718fd..9bf04de788 100644
> --- a/packages/js/src/components/WincherRankingHistoryChart.js
> +++ b/packages/js/src/components/WincherRankingHistoryChart.js
> @@ -27,8 +27,8 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
>                         y: value,
>                 } ) ),
>                 lineTension: 0,
> -               pointRadius: 0,
> -               pointHoverRadius: 2,
> +               pointRadius: 1,
> +               pointHoverRadius: 4,
>                 borderWidth: 2,
>         } ) );
> 
> @@ -58,7 +58,8 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
>                                                 },
>                                                 titleAlign: "center",
>                                                 intersect: false,
> -                                               mode: "nearest",
> +                                               mode: "point",
> +                                               position: "nearest",
>                                         },
>                                 },
>                                 scales: {
> ```

I'll check it. Meanwhile, could you please check this (maybe we can lower hitbox):

```diff
diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
index e4895718fd..b2f2347182 100644
--- a/packages/js/src/components/WincherRankingHistoryChart.js
+++ b/packages/js/src/components/WincherRankingHistoryChart.js
@@ -28,8 +28,9 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
                } ) ),
                lineTension: 0,
                pointRadius: 0,
-               pointHoverRadius: 2,
+               pointHoverRadius: 3,
                borderWidth: 2,
+               pointHitRadius: 10,
        } ) );

        return (
@@ -58,9 +59,17 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
                                                },
                                                titleAlign: "center",
                                                intersect: false,
-                                               mode: "nearest",
+                                               mode: "x",
                                        },
                                },
+                               hover: {
+                                       intersect: false,
+                                       mode: "x",
+                               },
+                               interaction: {
+                                  intersect: false,
+                                  mode: "x",
+                               },
                                scales: {
                                        x: {
                                                bounds: "ticks",
```

skrivanos commented:
> > > > @KaisZaoualiWincher Is tooltip shared between series? I don't have yoast premium atm so can't test
> > > 
> > > 
> > > It's not shared since we need `mode = "nearest"` otherwise it'll be hard to trigger the tooltip. I'd like to have a mix of "nearest" and "x" modes but we don't have that.
> > 
> > 
> > IMO this is better even though it's not ideal.
> > ```diff
> > diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
> > index e4895718fd..9bf04de788 100644
> > --- a/packages/js/src/components/WincherRankingHistoryChart.js
> > +++ b/packages/js/src/components/WincherRankingHistoryChart.js
> > @@ -27,8 +27,8 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
> >                         y: value,
> >                 } ) ),
> >                 lineTension: 0,
> > -               pointRadius: 0,
> > -               pointHoverRadius: 2,
> > +               pointRadius: 1,
> > +               pointHoverRadius: 4,
> >                 borderWidth: 2,
> >         } ) );
> > 
> > @@ -58,7 +58,8 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
> >                                                 },
> >                                                 titleAlign: "center",
> >                                                 intersect: false,
> > -                                               mode: "nearest",
> > +                                               mode: "point",
> > +                                               position: "nearest",
> >                                         },
> >                                 },
> >                                 scales: {
> > ```
> 
> I'll check it. Meanwhile, could you please check this (maybe we can lower hitbox):
> 
> ```diff
> diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
> index e4895718fd..b2f2347182 100644
> --- a/packages/js/src/components/WincherRankingHistoryChart.js
> +++ b/packages/js/src/components/WincherRankingHistoryChart.js
> @@ -28,8 +28,9 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
>                 } ) ),
>                 lineTension: 0,
>                 pointRadius: 0,
> -               pointHoverRadius: 2,
> +               pointHoverRadius: 3,
>                 borderWidth: 2,
> +               pointHitRadius: 10,
>         } ) );
> 
>         return (
> @@ -58,9 +59,17 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
>                                                 },
>                                                 titleAlign: "center",
>                                                 intersect: false,
> -                                               mode: "nearest",
> +                                               mode: "x",
>                                         },
>                                 },
> +                               hover: {
> +                                       intersect: false,
> +                                       mode: "x",
> +                               },
> +                               interaction: {
> +                                  intersect: false,
> +                                  mode: "x",
> +                               },
>                                 scales: {
>                                         x: {
>                                                 bounds: "ticks",
> ```

Seems to work decent. Although we have a problem. Sometimes you see multiple dates at once:

<img width="179" alt="Screenshot 2023-07-20 at 13 28 32" src="https://github.com/wincher-ab/wordpress-seo/assets/87463/85eaf086-cb9a-4dda-8358-da35b0546ebd">

KaisZaoualiWincher commented:
Yes, and the problem raises more when we have datapoints close to each other, even lowering the hitbox won't solve it..

skrivanos commented:
> Yes, and the problem raises more when we have datapoints close to each other, even lowering the hitbox won't solve it..

I feel like using mode point like what I did with a larger hitbox might be the best way.

KaisZaoualiWincher commented:
> > Yes, and the problem raises more when we have datapoints close to each other, even lowering the hitbox won't solve it..
> 
> I feel like using mode point like what I did with a larger hitbox might be the best way.

Let's try this, please? (maybe remove `position nearest` to show tooltip in average distance against series)

```diff
diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
index e4895718fd..2955278b26 100644
--- a/packages/js/src/components/WincherRankingHistoryChart.js
+++ b/packages/js/src/components/WincherRankingHistoryChart.js
@@ -28,8 +28,9 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
                } ) ),
                lineTension: 0,
                pointRadius: 0,
-               pointHoverRadius: 2,
+               pointHoverRadius: 3,
                borderWidth: 2,
+               pointHitRadius: 4,
        } ) );

        return (
@@ -59,8 +60,15 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
                                                titleAlign: "center",
                                                intersect: false,
                                                mode: "nearest",
+                                               position: "nearest",
+                                               axis: "x",
                                        },
                                },
+                               hover: {
+                                       intersect: false,
+                                       mode: "nearest",
+                                       axis: "x",
+                               },
                                scales: {
                                        x: {
                                                bounds: "ticks",
```

skrivanos commented:
> > > Yes, and the problem raises more when we have datapoints close to each other, even lowering the hitbox won't solve it..
> > 
> > 
> > I feel like using mode point like what I did with a larger hitbox might be the best way.
> 
> Let's try this, please? (maybe remove `position nearest` to show tooltip in average distance against series)
> 
> ```diff
> diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
> index e4895718fd..2955278b26 100644
> --- a/packages/js/src/components/WincherRankingHistoryChart.js
> +++ b/packages/js/src/components/WincherRankingHistoryChart.js
> @@ -28,8 +28,9 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
>                 } ) ),
>                 lineTension: 0,
>                 pointRadius: 0,
> -               pointHoverRadius: 2,
> +               pointHoverRadius: 3,
>                 borderWidth: 2,
> +               pointHitRadius: 4,
>         } ) );
> 
>         return (
> @@ -59,8 +60,15 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
>                                                 titleAlign: "center",
>                                                 intersect: false,
>                                                 mode: "nearest",
> +                                               position: "nearest",
> +                                               axis: "x",
>                                         },
>                                 },
> +                               hover: {
> +                                       intersect: false,
> +                                       mode: "nearest",
> +                                       axis: "x",
> +                               },
>                                 scales: {
>                                         x: {
>                                                 bounds: "ticks",
> ```

I don't like it much, it's sort of confusing how to get certain points tooltip to show. 

I don't really see why we're struggling so much instead of using mode point with a larger hit radius anyway.

KaisZaoualiWincher commented:
> > > > Yes, and the problem raises more when we have datapoints close to each other, even lowering the hitbox won't solve it..
> > > 
> > > 
> > > I feel like using mode point like what I did with a larger hitbox might be the best way.
> > 
> > 
> > Let's try this, please? (maybe remove `position nearest` to show tooltip in average distance against series)
> > ```diff
> > diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
> > index e4895718fd..2955278b26 100644
> > --- a/packages/js/src/components/WincherRankingHistoryChart.js
> > +++ b/packages/js/src/components/WincherRankingHistoryChart.js
> > @@ -28,8 +28,9 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
> >                 } ) ),
> >                 lineTension: 0,
> >                 pointRadius: 0,
> > -               pointHoverRadius: 2,
> > +               pointHoverRadius: 3,
> >                 borderWidth: 2,
> > +               pointHitRadius: 4,
> >         } ) );
> > 
> >         return (
> > @@ -59,8 +60,15 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
> >                                                 titleAlign: "center",
> >                                                 intersect: false,
> >                                                 mode: "nearest",
> > +                                               position: "nearest",
> > +                                               axis: "x",
> >                                         },
> >                                 },
> > +                               hover: {
> > +                                       intersect: false,
> > +                                       mode: "nearest",
> > +                                       axis: "x",
> > +                               },
> >                                 scales: {
> >                                         x: {
> >                                                 bounds: "ticks",
> > ```
> 
> I don't like it much, it's sort of confusing how to get certain points tooltip to show.
> 
> I don't really see why we're struggling so much instead of using mode point with a larger hit radius anyway.

No, it's just that I found the new combination (`mode nearest` and `axis x`) displaying the right x-axis values without confusion that we had experienced in the first patch (using only `mode nearest`).

https://watch.screencastify.com/v/Z3j6PUbDPeOZMCShquqS

But, yes, let's keep it simple as we do always. :)
Agreed, point mode and hitbox of 6 are perfect.

skrivanos commented:
> > > > > Yes, and the problem raises more when we have datapoints close to each other, even lowering the hitbox won't solve it..
> > > > 
> > > > 
> > > > I feel like using mode point like what I did with a larger hitbox might be the best way.
> > > 
> > > 
> > > Let's try this, please? (maybe remove `position nearest` to show tooltip in average distance against series)
> > > ```diff
> > > diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
> > > index e4895718fd..2955278b26 100644
> > > --- a/packages/js/src/components/WincherRankingHistoryChart.js
> > > +++ b/packages/js/src/components/WincherRankingHistoryChart.js
> > > @@ -28,8 +28,9 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
> > >                 } ) ),
> > >                 lineTension: 0,
> > >                 pointRadius: 0,
> > > -               pointHoverRadius: 2,
> > > +               pointHoverRadius: 3,
> > >                 borderWidth: 2,
> > > +               pointHitRadius: 4,
> > >         } ) );
> > > 
> > >         return (
> > > @@ -59,8 +60,15 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
> > >                                                 titleAlign: "center",
> > >                                                 intersect: false,
> > >                                                 mode: "nearest",
> > > +                                               position: "nearest",
> > > +                                               axis: "x",
> > >                                         },
> > >                                 },
> > > +                               hover: {
> > > +                                       intersect: false,
> > > +                                       mode: "nearest",
> > > +                                       axis: "x",
> > > +                               },
> > >                                 scales: {
> > >                                         x: {
> > >                                                 bounds: "ticks",
> > > ```
> > 
> > 
> > I don't like it much, it's sort of confusing how to get certain points tooltip to show.
> > I don't really see why we're struggling so much instead of using mode point with a larger hit radius anyway.
> 
> No, it's just that I found the new combination (`mode nearest` and `axis x`) displaying the right x-axis values without confusion that we had experienced in the first patch (using only `mode nearest`).
> 
> https://watch.screencastify.com/v/Z3j6PUbDPeOZMCShquqS
> 
> But, yes, let's keep it simple as we do always. :) Agreed, point mode and hitbox of 6 are perfect.

IDK why but I don't see the same results with this as your video. I get only one series in the tooltip.

KaisZaoualiWincher commented:
> > > > > > Yes, and the problem raises more when we have datapoints close to each other, even lowering the hitbox won't solve it..
> > > > > 
> > > > > 
> > > > > I feel like using mode point like what I did with a larger hitbox might be the best way.
> > > > 
> > > > 
> > > > Let's try this, please? (maybe remove `position nearest` to show tooltip in average distance against series)
> > > > ```diff
> > > > diff --git a/packages/js/src/components/WincherRankingHistoryChart.js b/packages/js/src/components/WincherRankingHistoryChart.js
> > > > index e4895718fd..2955278b26 100644
> > > > --- a/packages/js/src/components/WincherRankingHistoryChart.js
> > > > +++ b/packages/js/src/components/WincherRankingHistoryChart.js
> > > > @@ -28,8 +28,9 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
> > > >                 } ) ),
> > > >                 lineTension: 0,
> > > >                 pointRadius: 0,
> > > > -               pointHoverRadius: 2,
> > > > +               pointHoverRadius: 3,
> > > >                 borderWidth: 2,
> > > > +               pointHitRadius: 4,
> > > >         } ) );
> > > > 
> > > >         return (
> > > > @@ -59,8 +60,15 @@ export default function WincherRankingHistoryChart( { datasets, isChartShown } )
> > > >                                                 titleAlign: "center",
> > > >                                                 intersect: false,
> > > >                                                 mode: "nearest",
> > > > +                                               position: "nearest",
> > > > +                                               axis: "x",
> > > >                                         },
> > > >                                 },
> > > > +                               hover: {
> > > > +                                       intersect: false,
> > > > +                                       mode: "nearest",
> > > > +                                       axis: "x",
> > > > +                               },
> > > >                                 scales: {
> > > >                                         x: {
> > > >                                                 bounds: "ticks",
> > > > ```
> > > 
> > > 
> > > I don't like it much, it's sort of confusing how to get certain points tooltip to show.
> > > I don't really see why we're struggling so much instead of using mode point with a larger hit radius anyway.
> > 
> > 
> > No, it's just that I found the new combination (`mode nearest` and `axis x`) displaying the right x-axis values without confusion that we had experienced in the first patch (using only `mode nearest`).
> > https://watch.screencastify.com/v/Z3j6PUbDPeOZMCShquqS
> > But, yes, let's keep it simple as we do always. :) Agreed, point mode and hitbox of 6 are perfect.
> 
> IDK why but I don't see the same results with this as your video. I get only one series in the tooltip.

I have no idea :(

I have also fixed the case where we have more than one single position for a date.

bug:

![image](https://github.com/wincher-ab/wordpress-seo/assets/129404933/6b6d108e-849d-48f6-8b9c-079de80af275)

fixed:

![image](https://github.com/wincher-ab/wordpress-seo/assets/129404933/6ce8d42e-be53-4743-9051-d108e1a88a43)

I have made the link series in a separate commit, if we do not agree about it, let's just drop it. :)

skrivanos commented:
Seems broken to me, look at this. I can't see 2 series ever, plus it's sorta random which one is shown. IMO how it is now is not better than point mode.


https://github.com/wincher-ab/wordpress-seo/assets/87463/cb0ff69e-87f6-488a-91bf-7da0e36bbb56

KaisZaoualiWincher commented:
> Seems broken to me, look at this. I can't see 2 series ever, plus it's sorta random which one is shown. IMO how it is now is not better than point mode.
> 
>  Screen.Recording.2023-07-20.at.18.26.02.mov

That's so weird since it's working fine for me. OK, let's drop the last commit then.

KaisZaoualiWincher commented:
> Seems broken to me, look at this. I can't see 2 series ever, plus it's sorta random which one is shown. IMO how it is now is not better than point mode.
> 
> Screen Recording 2023-07-20 at 18.26.02.mov

Set it back to point mode

skrivanos commented:
> > Seems broken to me, look at this. I can't see 2 series ever, plus it's sorta random which one is shown. IMO how it is now is not better than point mode.
> > Screen Recording 2023-07-20 at 18.26.02.mov
> 
> Set it back to point mode

Works good now.

KaisZaoualiWincher commented:
@skrivanos new comments should be addressed!

skrivanos commented on packages/js/src/components/WincherTableRow.js:
```diff
@@ -132,6 +132,15 @@ export function getKeyphrasePosition( keyphrase ) {
+ *
+ * @returns {string} The formatted last updated date.
+ */
+const formatLastUpdated = ( dateString ) => moment( dateString ).fromNow();
```
I realized... Will this work even for other languages? Not sure on Yoast's policy on translations nor how it works with moment. Maybe it's auto-handled somehow.

KaisZaoualiWincher replied:
Seems to work fine and is related to wordpress admin language (settings --> General)

![image](https://github.com/wincher-ab/wordpress-seo/assets/129404933/ce842b4d-4598-48ec-96c3-a7a44259ed85)

skrivanos replied:
That's awesome!

skrivanos commented:
Ok I guess we're ready to send to to the Yoast team for their review! GJ

KaisZaoualiWincher commented:
> Ok I guess we're ready to send to to the Yoast team for their review! GJ

PR on Yoast repo: https://github.com/Yoast/wordpress-seo/pull/20522

KaisZaoualiWincher committed:
Add horizontal margin to Wincher chart

KaisZaoualiWincher committed:
Restyle Wincher table

KaisZaoualiWincher committed:
Remove show/hide wincher chart button

KaisZaoualiWincher committed:
Add keyphrase to Wincher chart when tracked

KaisZaoualiWincher committed:
Show message for untracked keyphrases in Wincher table

KaisZaoualiWincher committed:
Sort Wincher table keyphrases by tracked

KaisZaoualiWincher committed:
Center tracking toggle in Wincher table row

KaisZaoualiWincher committed:
Use modal instead of metabox for Wincher SEO performance

KaisZaoualiWincher committed:
Fix focus keyphrase when there is no focus keyphrase

KaisZaoualiWincher committed:
Replace deprecated babel plugin proposal

KaisZaoualiWincher committed:
Fix last month period option when history limit is 31 days

KaisZaoualiWincher committed:
Remove check limit call for not logged in users

KaisZaoualiWincher committed:
Use statefull collapsible instead of stateless one in WincherSEOPerformanceModal

KaisZaoualiWincher committed:
Use utc for start at date

KaisZaoualiWincher committed:
Fix open modal in elementor

KaisZaoualiWincher committed:
Fix last year is shown when it should not

KaisZaoualiWincher committed:
Fix dataset color changing when deselected then selected in Wincher chart

KaisZaoualiWincher committed:
Do not display previous post keyphrases in Wincher table

KaisZaoualiWincher committed:
Fix tracking keyword is crashing Wincher chart
